lib/rust/profiler/flame-graph/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+//! This module contains functionality that allows the profiling framework to
+//! generate the data required to render a flame graph. This means creating data for each block
+//! that is supposed to be rendered, with start time, end time and labels.
+
+
+use enso_profiler as profiler;
+use enso_profiler_data as data;
+use enso_profiler_data::Interval;
+use enso_profiler_data::Lifetime;
+use enso_profiler_data::Measurement;
+
+
+// ==================
+// === Block Data ===
+// ==================
+
+/// A `Block` contains the data required to render a single block of a frame graph.
+#[derive(Clone, Debug)]
+pub struct Block {
+    /// Start x coordinate of the block.
+    pub start: f64,
+    /// End x coordinate of the block.
+    pub end:   f64,
+    /// Row that the block should be placed in.
+    pub row:   u32,
+    /// The label to be displayed with the block.
+    pub label: String,
+}
+
+impl Block {
+    /// Width of the block.
+    pub fn width(&self) -> f64 {
+        self.end - self.start
+    }
+}
+
+
+
+// ========================
+// === Flame Graph Data ===
+// ========================
+
+/// A `FlameGraph`, contains the information required to render a flame graph, i.e., the data for
+/// all blocks that make up the flame graph.
+#[derive(Debug, Default)]
+pub struct FlameGraph {
+    /// Collection of all blocks making up the flame graph.
+    pub blocks: Vec<Block>,
+}
+
+fn blocks_from_measurement<Metadata>(measurement: &Measurement<Metadata>, row: u32) -> Vec<Block> {
```

Comment:
This will not produce an icicle graph, because an async task is not active during the intervals its children are.

For some async tasks **a**, **b**, **c**, **d**; where:
a.children = [b, c]
c.children = [d]

This would make a graph like this:

```
aa   aaa      a
  bb    cc  bb
          dd
```

(Sync subtrees would be stacked right below each interval).

---

To get an icicle graph, you'd squish everything to the same baseline (like in Tetris, but going up the page):

```aabbaaaccddbba```

(Sync subtrees would all start at row 1, or row 0 when they ever occur without an async parent).

---

This could be implemented by: Non-async tasks go in row `parent_row+1`; *all async tasks go in row 0*.

(This works because an async task never has an active parent, which can be derived from two properties:
- There is never more than one async task active at a time (by nature of async), so an async task can't have an active async ancestor.
- If an async task has a non-async parent, the parent started it with `spawn`[^1]; in this case, the parent will finish before the executor gets a chance to start the child.)

[^1]: the alternative, `block_on` would be a different story, but we don't use it outside tests and surely never will; it's extremely hazardous for performance.

Reply:
Good point. I changed the code to align the async blocks in row 0. This seems to work fine for my test.